src/test/java/net/snowflake/client/jdbc/ResultSetLatestIT.java

```diff
@@ -120,6 +120,34 @@ public void testChunkDownloaderNoHang() throws SQLException {
     connection.close();
   }
 
+  /** This tests that the SnowflakeChunkDownloader doesn't hang when memory limits are low. */
+  @Test
+  public void testChunkDownloaderSetRetry() throws SQLException {
+    int stmtCount = 3;
+    int rowCount = 170000;
+    Connection connection = getConnection();
+    connection
+        .unwrap(SnowflakeConnectionV1.class)
+        .getSFBaseSession()
+        .setMemoryLimitForTesting(2000000);
+    // Set memory limit to low number
+    // open multiple statements concurrently to overwhelm current memory allocation
+    for (int i = 0; i < stmtCount; ++i) {
+      long start = System.currentTimeMillis();
+      Statement stmt = connection.createStatement();
+      ResultSet resultSet =
+          stmt.executeQuery(
+              "select randstr(100, random()) from table(generator(rowcount => " + rowCount + "))");
+      // consume half of the results and go to the next statement
+      for (int j = 0; j < rowCount / 2; j++) {
+        resultSet.next();
+      }
```

Comment:
Nice test. So what is the behavior in this test without the fix above ? It just hangs ?

Reply:
Yes

Reply:
It will hit the bug described in the description of this PR. It will wait on a condition with a time out of 1 hour.

Reply:
I agree the waiting time is too long but if no waiting at all, would it triggers OOM instead in reality?

Reply:
JDBC always handles OOM error, so user will not see raw OOM error messages.
After this fix, JDBC will throw this error after retrying for 10 times, but still cannot allocate enough memory to download the chunk.
```
downloader encountered error: Max retry reached for the download of #chunk37 (Total chunks: 108) retry=3, error=net.snowflake.client.jdbc.SnowflakeSQLLoggedException: JDBC driver internal error: Exception: Failure allocating buffer..
Caused by: org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 16777216 byte(s) of direct memory (used: 100663296, max: 100663296)
```
While previously when JDBC reaches prefetch memory limit (1.5G) it will wait for 1 hour, then throw this error:
```
Timeout waiting for the download of #chunk37 (Total chunks: 108) retry=3
```
I think after the fix, the error is more clear and user can take action based on it.

Reply:
Nice! Thanks for the explanation!